incomplete_cooperative/regret.py

```diff
@@ -0,0 +1,214 @@
+    To achieve this, we use a map from the coalitions (excluding K_0) to their "player_ids".
+
+    We minimize regret in every internal node of the game tree.
+    When updating regret, we need to scan through the meta-coalitions from largest to smallest.
+    This identification is refered to as the "rank" of the meta-coalition.
+    Such an order is found using the `metacoalition_ids_by_coalition_size` function, and stored
+    in `meta_id_to_rank`.
+
+    Attributes:
+        number_of_players: number of players in the games.
+        number_of_coalitions: number of coalitions which aren't in K_0.
+        limit_of_revealed: maximum number of coalitions to reveal, i.e. maximum size of metacoal.
+        meta_rank_to_id: map metacoalition size rank -> metacoalition id.
+        meta_id_to_rank: map metacoalition id -> metacoalition size rank.
+        cumulative_regret: an array of instant regrets of each regret minimizer (zero if an action is inapplicable).
+            This array has size (number_of_regret_minimizers, number_of_coalitions).
+            It is indexed by the meta-coalition ranks.
+    """
+
+    def __init__(self, number_of_players: int, limit_of_revealed: int, plus: bool = False) -> None:
+        """Initialize RM.
+
+        Arguments:
+            number_of_players: number of players in the games.
+            limit_of_revealed: maximum number of coalitions to reveal.
+            plus: whether to compute regret matching plus. Default: False.
+        """
+        self.plus = plus
+        self.number_of_players = number_of_players
+        # number of coalitions we are interested in -- ie not those in K_0
+        self.number_of_coalitions = 2 ** number_of_players - number_of_players - 2
+        self.limit_of_revealed = limit_of_revealed
+
+        # map metacoalition size rank -> metacoalition id
+        self.meta_rank_to_id = metacoalition_ids_by_coalition_size(number_of_players, limit_of_revealed)
+        # these are the metacoalitions of size <= limit_of_revealed
+        self.viable_metacoalitions = len(self.meta_rank_to_id)
+        # map metacoalition id -> metacoalition size rank
+        self.meta_id_to_rank = np.zeros(self.viable_metacoalitions, dtype=int)
+        self.meta_id_to_rank[self.meta_rank_to_id] = np.arange(self.viable_metacoalitions)
+
+        # regret minimizer for every internal node -- except for the bottom layer of game tree
+        self.number_of_regret_minimizers = coalitions_up_to(self.number_of_coalitions, limit_of_revealed - 1)
+
+        self.coalitions_to_player_ids = get_coalition_player_id_map(number_of_players)
+        self.cumulative_regret = np.zeros((self.number_of_regret_minimizers, self.number_of_coalitions), dtype=RMValue)
+        self.cumulative_strategy = np.zeros((self.number_of_regret_minimizers, self.number_of_coalitions),
+                                            dtype=RMValue)
+        self.iteration = 0
+
+    def save(self, path: Path) -> None:
```

Comment:
Finally implemented save and load methods. I decided to store the np arrays using `np.save`, as I think it might be more optimized. The rest is stored as a json file next to the stored np arrays. This means that it doesn't create one file, but rather a directory of 3 files, but I don't think that is a problem. What do you think?

Reply:
For 2D arrays, I think `np.savetxt` uses less memory. But it should be fine regardless. Feel free to resolve this conversation.

Reply:
Good to know. Leaving it as is now, and if it is horrendous, I will change it.